README
CHI-Driver-Memcached

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc CHI::Driver::Memcached

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=CHI-Driver-Memcached

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/CHI-Driver-Memcached

    CPAN Ratings
        http://cpanratings.perl.org/d/CHI-Driver-Memcached

    Search CPAN
        http://search.cpan.org/dist/CHI-Driver-Memcached


COPYRIGHT AND LICENCE

Copyright (C) 2008 Jonathan Swartz

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for CHI-Driver-Memcached

** denotes an incompatible change

0.16  Jan 26, 2014

0.15  Jan 26, 2014

- replace deprecated Class::MOP::class_load with Class::Load::class_load

0.14  Nov 30, 2011

- allow max_key_length to be overriden

0.13  Apr 18, 2011

- indicate support for expires_on_backend
- migrate to Dist::Zilla

0.12  Jun 9, 2010

- define max_key_length for auto-digesting in CHI >0.35
- skip tests for new test keys in CHI >0.35 containing control characters or whitespace

0.11  Jan 22, 2010

- skip tests if particular memcached library not available - reported by David Yee

0.10  Nov 29, 2009

- support CHI::Driver::Memcached::Fast and CHI::Driver::Memcached::libmemcached
- allow memd_params to be passed explicitly
- depend on CHI 0.33 for non_common_constructor_params fix

0.09  Nov 23, 2009

- fix failing tests
- depend on CHI 0.31

0.08  May 27, 2009

- depend on CHI 0.241, for fuller set of dependencies

0.07  May 26, 2009

- Revert test script to previous format

0.06  May 25, 2009

- Switch back to Moose, to match CHI 0.23
- Extend list of keys to check in tests, to match CHI 0.23

0.05  Mar 2, 2009

- Switch to Mouse, to match CHI 0.10

0.04  Dec 30, 2008

- Moved source to git
- Append colon to namespace before passing to Cache::Memcached

0.03  Sep 26, 2008

- Require Cache::Memcached
- Skip test if memcached not running

0.02  Sep 24, 2008

- Require CHI 0.081 - will not work with previous versions

0.01  Sep 23, 2008

- Initial version


lib/CHI/Driver/Memcached/Test.pm
# $Id: $
#
package CHI::Driver::Memcached::Test;
use CHI::Driver::Memcached::Util qw(require_dynamic);
use strict;
use warnings;

sub import {
    my $class = shift;
    $class->export_to_level( 1, undef, @_ );
}

sub packages_to_import {
    return (
        qw(
          Test::Deep
          Test::More
          Test::Exception
          CHI::Driver::Memcached::Test::Util
          )
    );
}

sub export_to_level {
    my ( $class, $level, $ignore ) = @_;

    foreach my $package ( $class->packages_to_import() ) {
        require_dynamic($package);
        my @export;
        if ( $package eq 'Test::Deep' ) {

            # Test::Deep exports way too much by default
            @export =
              qw(eq_deeply cmp_deeply cmp_set cmp_bag cmp_methods subbagof superbagof subsetof supersetof superhashof subhashof);
        }
        else {

            # Otherwise, grab everything from @EXPORT
            no strict 'refs';
            @export = @{"$package\::EXPORT"};
        }
        $package->export_to_level( $level + 1, undef, @export );
    }
}

1;


lib/CHI/Driver/Memcached/t/Sanity.pm
package CHI::Driver::Memcached::t::Sanity;
use strict;
use warnings;
use CHI::Driver::Memcached::Test;
use base qw(CHI::Driver::Memcached::Test::Class);

sub test_ok : Test(1) {
    ok( 1, '1 is ok' );
}

1;


lib/CHI/Driver/Memcached/Test/Util.pm
package CHI::Driver::Memcached::Test::Util;
use Test::Builder;
use Test::More;
use strict;
use warnings;
use base qw(Exporter);

our @EXPORT_OK = qw();

1;


lib/CHI/Driver/Memcached/Base.pm
package CHI::Driver::Memcached::Base;

use CHI;
use Carp;
use Class::Load;
use Moose;
use strict;
use warnings;

has 'memd'        => ( is => 'ro', init_arg => undef );
has 'memd_class'  => ( is => 'ro' );
has 'memd_params' => ( is => 'ro' );

extends 'CHI::Driver::Base::CacheContainer';

# Unsupported methods
#
__PACKAGE__->declare_unsupported_methods(
    qw(dump_as_hash get_keys get_namespaces is_empty clear purge));

__PACKAGE__->meta->make_immutable();

sub BUILD {
    my ( $self, $params ) = @_;

    $self->{memd_params} ||= $self->non_common_constructor_params($params);
    $self->{memd_params}->{namespace} ||= $self->{namespace} . ":";
    $self->{memd} = $self->{_contained_cache} = $self->_build_contained_cache;
    $self->{max_key_length} = 248 - length( $self->{namespace} )
      if !defined( $self->{max_key_length} );
}

sub _build_contained_cache {
    my ($self) = @_;

    Class::Load::load_class( $self->memd_class );
    return $self->memd_class->new( $self->memd_params );
}

# Memcached supports fast multiple get
#

sub fetch_multi_hashref {
    my ( $self, $keys ) = @_;
    croak "must specify keys" unless defined($keys);

    return $self->memd->get_multi(@$keys);
}

1;


lib/CHI/Driver/Memcached/Fast.pm
package CHI::Driver::Memcached::Fast;

use Moose;
use strict;
use warnings;

extends 'CHI::Driver::Memcached::Base';

has '+memd_class' => ( default => 'Cache::Memcached::Fast' );

__PACKAGE__->meta->make_immutable();

1;

__END__

=pod

=head1 NAME

CHI::Driver::Memcached::Fast -- Distributed cache via Cache::Memcached::Fast

=head1 SYNOPSIS

    use CHI;

    my $cache = CHI->new(
        driver => 'Memcached::Fast',
        namespace => 'products',
        servers => [ "10.0.0.15:11211", "10.0.0.15:11212", "/var/sock/memcached",
        "10.0.0.17:11211", [ "10.0.0.17:11211", 3 ] ],
        debug => 0,
        compress_threshold => 10_000,
    );

=head1 DESCRIPTION

A CHI driver that uses L<Cache::Memcached::Fast|Cache::Memcached::Fast> to
store data in the specified memcached server(s). From the perspective of the
CHI API, the feature set is nearly identical to
L<CHI::Driver::Memcached|CHI::Driver::Memcached>.

=head1 AUTHOR

Jonathan Swartz

=head1 SEE ALSO

L<CHI|CHI>, L<Cache::Memcached::Fast|Cache::Memcached::Fast>,
L<CHI::Driver::Memcached>, L<CHI::Driver::Memcached::libmemcached>

=head1 COPYRIGHT & LICENSE

Copyright (C) 2007 Jonathan Swartz.

CHI::Driver::Memcached is provided "as is" and without any express or implied
warranties, including, without limitation, the implied warranties of
merchantibility and fitness for a particular purpose.

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/CHI/Driver/Memcached/libmemcached.pm
package CHI::Driver::Memcached::libmemcached;

use Moose;
use strict;
use warnings;

extends 'CHI::Driver::Memcached::Base';

has '+memd_class' => ( default => 'Cache::Memcached::libmemcached' );

__PACKAGE__->meta->make_immutable();

1;

__END__

=pod

=head1 NAME

CHI::Driver::Memcached::libmemcached -- Distributed cache via
Cache::Memcached::libmemcached

=head1 SYNOPSIS

    use CHI;

    my $cache = CHI->new(
        driver => 'Memcached::libmemcached',
        namespace => 'products',
        servers => [ "10.0.0.15:11211", "10.0.0.15:11212", "/var/sock/memcached",
        "10.0.0.17:11211", [ "10.0.0.17:11211", 3 ] ],
        debug => 0,
        compress_threshold => 10_000,
    );

=head1 DESCRIPTION

A CHI driver that uses
L<Cache::Memcached::libmemcached|Cache::Memcached::libmemcached> to store data
in the specified memcached server(s). From the perspective of the CHI API, the
feature set is nearly identical to
L<CHI::Driver::Memcached|CHI::Driver::Memcached>.

=head1 AUTHOR

Jonathan Swartz

=head1 SEE ALSO

L<CHI|CHI>, L<Cache::Memcached::libmemcached|Cache::Memcached::libmemcached>,
L<CHI::Driver::Memcached>, L<CHI::Driver::Memcached::Fast>

=head1 COPYRIGHT & LICENSE

Copyright (C) 2007 Jonathan Swartz.

CHI::Driver::Memcached is provided "as is" and without any express or implied
warranties, including, without limitation, the implied warranties of
merchantibility and fitness for a particular purpose.

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/CHI/Driver/Memcached.pm
package CHI::Driver::Memcached;

use Moose;
use strict;
use warnings;

extends 'CHI::Driver::Memcached::Base';

has '+memd_class' => ( default => 'Cache::Memcached' );

__PACKAGE__->meta->make_immutable();

1;

__END__

=pod

=head1 NAME

CHI::Driver::Memcached -- Distributed cache via memcached (memory cache daemon)

=head1 SYNOPSIS

    use CHI;

    my $cache = CHI->new(
        driver => 'Memcached',   # or 'Memcached::Fast', or 'Memcached::libmemcached'
        namespace => 'products',
        servers => [ "10.0.0.15:11211", "10.0.0.15:11212", "/var/sock/memcached",
        "10.0.0.17:11211", [ "10.0.0.17:11211", 3 ] ],
        debug => 0,
        compress_threshold => 10_000,
    );

=head1 DESCRIPTION

A CHI driver that uses Cache::Memcached to store data in the specified
memcached server(s).

L<CHI::Driver::Memcached::Fast> and L<CHI::Driver::Memcached::libmemcached> are
also available as part of this distribution. They work with other Memcached
clients and support a similar feature set. Documentation for all three modules
is presented below.

=head1 CONSTRUCTOR OPTIONS

Namespace, appended with ":", is passed along to the Cached::Memcached::*
constructor, along with any constructor options L<not recognized by
CHI|CHI/constructor> - for example I<servers>, I<compress_threshold> and
I<debug>.

If you need more control over the options passed to Cache::Memcached::*, you
may specify a hash directly in C<memd_params>.

=head1 METHODS

Besides the standard CHI methods:

=over

=item memd

Returns a handle to the underlying Cache::Memcached::* object. You can use this
to call memcached-specific methods that are not supported by the general API,
e.g.

    $self->memd->incr("key");
    my $stats = $self->memd->stats();

=back

=head1 UNSUPPORTED METHODS

These standard CHI methods cannot currently be supported by memcached, chiefly
because there is no way to get a list of stored keys.

=over

=item dump_as_hash

=item clear

=item get_keys

=item get_namespaces

=item is_empty

=item purge

=back

=head1 SUPPORT AND DOCUMENTATION

Questions and feedback are welcome, and should be directed to the perl-cache
mailing list:

    http://groups.google.com/group/perl-cache-discuss

Bugs and feature requests will be tracked at RT:

    http://rt.cpan.org/NoAuth/Bugs.html?Dist=CHI-Driver-Memcached

The latest source code can be browsed and fetched at:

    http://github.com/jonswar/perl-chi-driver-memcached/tree/master
    git clone git://github.com/jonswar/perl-chi-driver-memcached.git

=head1 AUTHOR

Jonathan Swartz

=head1 SEE ALSO

L<CHI|CHI>, L<Cache::Memcached|Cache::Memcached>,
L<CHI::Driver::Memcached::Fast>, L<CHI::Driver::Memcached::libmemcached>

=head1 COPYRIGHT & LICENSE

Copyright (C) 2007 Jonathan Swartz.

CHI::Driver::Memcached is provided "as is" and without any express or implied
warranties, including, without limitation, the implied warranties of
merchantibility and fitness for a particular purpose.

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
